server/src/main/java/org/opensearch/common/breaker/ResponseLimitSettings.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.breaker;
+
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.IndexRoutingTable;
+import org.opensearch.cluster.routing.IndexShardRoutingTable;
+import org.opensearch.cluster.routing.RoutingTable;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.rest.action.cat.RestIndicesAction;
+import org.opensearch.rest.action.cat.RestSegmentsAction;
+import org.opensearch.rest.action.cat.RestShardsAction;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Class to define dynamic settings for putting response limits on the actions and methods to evaluate if block is required.
+ */
+public class ResponseLimitSettings {
+
+    /**
+     * Enum to represent entity against which we need to perform limit checks.
+     */
+    public enum LimitEntity {
+        INDICES,
+        SHARDS
+    }
+
+    private volatile int catIndicesResponseLimit;
+    private volatile int catShardsResponseLimit;
+    private volatile int catSegmentsResponseLimit;
+
+    /**
+     * Setting to enable response limit on {@link RestIndicesAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_INDICES_RESPONSE_LIMIT_SETTING = Setting.intSetting(
+        "cat.indices.response.limit.number_of_indices",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable response limit on {@link RestShardsAction}. The limit will be applied on number of shards.
+     */
+    public static final Setting<Integer> CAT_SHARDS_RESPONSE_LIMIT_SETTING = Setting.intSetting(
+        "cat.shards.response.limit.number_of_shards",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable response limit on {@link RestSegmentsAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_SEGMENTS_RESPONSE_LIMIT_SETTING = Setting.intSetting(
+        "cat.segments.response.limit.number_of_indices",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
```

Comment:
It might be good to define a lower and upper bound value for these settings. However this can be taken as a follow-up as seperate PR.